### PR TITLE
Add CodeChecker to CI

### PR DESCRIPTION
**Pull request by W-Mai**
## Summary

Add CodeChecker to CI

## Impact

Increase the rigor with which code is reviewed
It may increase the time of CI

### CODE REVIEW

File: tools/ci/runcodechecker.sh [hunk 1]
```diff
@@ -93,3 +116,11 @@ done
...
   start_server
 fi
+
+if $NEEDKILL; then
+  kill_server
+fi
+
+if $NEEDCOMPRESS; then
+  compress_database $NEEDRM
+fi
```
Comment (pkarashchenko) [inline tools/ci/runcodechecker.sh:126]:
```suggestion
fi

```

**Comment by xiaoxiang781216** [comment]
@W-Mai please squash your change into one patch

File: tools/ci/runcodechecker.sh [hunk 2]
```diff
@@ -0,0 +1,164 @@
+#!/usr/bin/env bash
+# tools/ci/runcodechecker.sh
```
Comment (xiaoxiang781216) [inline tools/ci/runcodechecker.sh:2]: I don't see any place invoke this shell script

Comment (W-Mai) [reply tools/ci/runcodechecker.sh:2]: https://github.com/apache/incubator-nuttx/blob/198e093ec2733dea7e818d84f346369cf7902f9e/.github/workflows/build.yml#L166-L175

It is called in `.github/workflows/build.yml`

**Comment by xiaoxiang781216** [comment]
@W-Mai please remove the merge patch.

**Comment by W-Mai** [comment]
> @W-Mai please remove the merge patch.

I'll rebase it later.

File: tools/testbuild.sh [hunk 3]
```diff
@@ -208,13 +221,63 @@ function exportandimport {
...
+
+  pushd $target_path >/dev/null
+
+  tar zcf ${target_name}.tar.gz ${target_name}
+  rm -rf ${target_name} 
+
+  popd >/dev/null
+}
+
 function makefunc {
```
Comment (xiaoxiang781216) [inline tools/testbuild.sh:236]: let's add checkfunc and call it from build function

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -157,17 +158,35 @@ jobs:
             export CCACHE_DIR=`pwd`/ccache
             mkdir $CCACHE_DIR
             export ARTIFACTDIR=`pwd`/buildartifacts
+            export CODECHECKERLOGSDIR=`pwd`/codecheckerlogs
```
[the remark this answers is not in the document]
Comment (W-Mai) [reply .github/workflows/build.yml:161]: When the build fails, the upload step is skipped

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -157,17 +158,35 @@ jobs:
...
+            fi
             ccache -s
 
       - uses: actions/upload-artifact@v3
         with:
           name: linux-builds
           path: buildartifacts/
         continue-on-error: true
+      
+      - name: Upload CodeChecker Logs
```
[the remark this answers is not in the document]
Comment (W-Mai) [reply .github/workflows/build.yml:183]: When the build fails, the buildartifacts step is skipped

[the remark this answers is not in the document]
Comment (W-Mai) [reply .github/workflows/build.yml:183]: But we may need it to upload `html report` generated by CodeChecker.

File: tools/testbuild.sh [hunk 6]
```diff
@@ -57,7 +62,7 @@ esac
 
 function showusage {
   echo ""
-  echo "USAGE: $progname [-l|m|c|g|n] [-d] [-e <extraflags>] [-x] [-j <ncpus>] [-a <appsdir>] [-t <topdir>] [-p] [-G] <testlist-file>"
+  echo "USAGE: $progname [-l|m|c|g|n] [-d] [-e <extraflags>] [-x] [-j <ncpus>] [-a <appsdir>] [-t <topdir>] [-p] [-G] [--codechecker] [--codechecker-failed] <testlist-file>"
```
[the remark this answers is not in the document]
Comment (W-Mai) [reply tools/testbuild.sh:65]: that's ok

File: tools/testbuild.sh [hunk 7]
```diff
@@ -218,6 +236,32 @@ function makefunc {
...
+  codecheck_ret=$?
+  echo "    Storing analysis result to CodeChecker..."
+  echo "      Generating HTML report..."
+  CodeChecker parse --export html --output "${codechecker_logs_dir}/codechecker_reports_html" "${codechecker_logs_dir}/codechecker_reports" 1>/dev/null
+  echo "      Compressing logs..."
+  compressartifacts "$(dirname "${codechecker_logs_dir}")" "${sub_target_name}"
+
+# If you need to stop CI, uncomment the following line.
+#  if [ $codecheck_ret -ne 0 ]; then
+#    fail=114
```
Comment (xiaoxiang781216) [inline tools/testbuild.sh:259]: fail=1

File: tools/testbuild.sh [hunk 8]
```diff
@@ -36,6 +36,8 @@ PRINTLISTONLY=0
 GITCLEAN=0
 SAVEARTIFACTS=0
 CHECKCLEAN=1
+CODECHECKER=0
+CODECHECKERFAILS=0
```
Comment (xiaoxiang781216) [inline tools/testbuild.sh:40]: remove

File: tools/testbuild.sh [hunk 9]
```diff
@@ -218,6 +236,32 @@ function makefunc {
...
+  build_cmd="${MAKE} ${MAKE_FLAGS} \"${EXTRA_FLAGS}\" ${JOPTION} 1>/dev/null"
+
+  local config_sub_path=$(echo "$config" | sed "s/:/\//")
+  local sub_target_name=${config_sub_path#$(dirname "${config_sub_path}")/}
+  local codechecker_logs_dir=${ARTIFACTDIR}/codechecker_logs/${config_sub_path}
+
+  mkdir -p "${codechecker_logs_dir}"
+
+  echo "    Checking NuttX by Codechecker..."
+  CodeChecker check -b "${build_cmd}" -o "${codechecker_logs_dir}"/codechecker_reports -e sensitive --ctu
```
Comment (xiaoxiang781216) [inline tools/testbuild.sh:249]:
```suggestion
  CodeChecker check -b "${build_cmd}" -o "${codechecker_logs_dir}"/log -e sensitive --ctu
```

File: tools/testbuild.sh [hunk 10]
```diff
@@ -218,6 +236,32 @@ function makefunc {
...
+  codecheck_ret=$?
+  echo "    Storing analysis result to CodeChecker..."
+  echo "      Generating HTML report..."
+  CodeChecker parse --export html --output "${codechecker_logs_dir}/codechecker_reports_html" "${codechecker_logs_dir}/codechecker_reports" 1>/dev/null
```
Comment (xiaoxiang781216) [inline tools/testbuild.sh:253]:
```suggestion
  CodeChecker parse --export html --output "${codechecker_logs_dir}/html" "${codechecker_logs_dir}/codechecker_reports" 1>/dev/null
```

File: tools/testbuild.sh [hunk 11]
```diff
@@ -218,6 +236,32 @@ function makefunc {
   return $fail
 }
 
+function checkfunc {
+  build_cmd="${MAKE} ${MAKE_FLAGS} \"${EXTRA_FLAGS}\" ${JOPTION} 1>/dev/null"
+
+  local config_sub_path=$(echo "$config" | sed "s/:/\//")
+  local sub_target_name=${config_sub_path#$(dirname "${config_sub_path}")/}
+  local codechecker_logs_dir=${ARTIFACTDIR}/codechecker_logs/${config_sub_path}
```
Comment (xiaoxiang781216) [inline tools/testbuild.sh:244]:
```suggestion
  local codechecker_dir=${ARTIFACTDIR}/codechecker
```

Comment (xiaoxiang781216) [reply tools/testbuild.sh:244]: why need config_sub_path? let's remove it.

Comment (W-Mai) [reply tools/testbuild.sh:244]: If there are multiple targets in codechecker.dat, the results of the checking may make the folder look disordered.
Now it's a separate folder for each target.

Comment (xiaoxiang781216) [reply tools/testbuild.sh:243]: Ok, let's change codechecker_logs_dir to codechecker_dir

Comment (W-Mai) [reply tools/testbuild.sh:243]: that's ok

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by pkarashchenko — APPROVED** [review]

**Pull request merged**